File: android/app/src/main/java/miko/todayinspacehistory/data/network/Endpoints.kt
package miko.todayinspacehistory.data.network

import miko.todayinspacehistory.util.todayDayMonthComponents
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

enum class Endpoints(val path: String) {
    Search("/search"),
    Asset("/asset/"),
    Captions("/captions/"),
    Album("/album/");

    val queryParams: Map<String, String>
        get() = when (this) {
            Search -> mapOf(
                "description" to todayDayMonthComponents().joinToString(" "),
                "media_type" to "image",
            )
            Asset, Captions, Album -> emptyMap()
        }

    val url: String?
        get() = when (this) {
            Search -> {
                val query = queryParams.entries.joinToString("&") { (key, value) ->
                    val encoded = URLEncoder.encode(value, StandardCharsets.UTF_8)
                    "$key=$encoded"
                }
                "${ApiConstants.BASE_URL}$path?$query"
            }
            Asset, Captions, Album -> null
        }

    val method: HttpMethod
        get() = HttpMethod.GET
}

File: ios/TodayInSpaceHistory/Networking/Endpoints.swift
//
//  Endpoints.swift
//  TodayInSpaceHistory
//
//  Created by Mikołaj Linczewski on 26/09/2021.
//

import Foundation

enum Endpoints: String {
    case search = "/search"
    case asset = "/asset/"
    case captions = "/captions/"
    case album = "/album/"
    
    var queryParams: [URLQueryItem] {
        switch self {
        case .search:
            return [
                URLQueryItem(
                    name: "description",
                    value: Date.todayDayMonthComponents().joined(separator: " ")
                ),
                URLQueryItem(name: "media_type", value: "image")
            ]
        case .asset, .captions, .album:
            return []
        }
    }
    
    var url: URL? {
        switch self {
        case .search:
            var components = URLComponents(string: APIConstants.baseURL + rawValue)
            components?.queryItems = queryParams
            return components?.url
        case .asset, .captions, .album:
            return nil
        }
    }
    
    var method: HTTPMethods {
        switch self {
        case .search: return .get
        case .asset: return .get
        case .captions: return .get
        case .album: return .get
        }
    }
}
